--- 
title: "Applying DoRothEAâ€™s regulons on single-cell RNA-seq data"
author:
- name: Alberto Valdeolivas
  affiliation: Institute for Computational Biomedicine, Heidelberg University
  email: alberto.valdeolivas@bioquant.uni-heidelberg.de
- name: Minoo Ashtiani 
  affiliation: Institute for Computational Biomedicine, Heidelberg University
- name: Christian Holland 
  affiliation: Institute for Computational Biomedicine, Heidelberg University
- name: Julio Saez-Rodriguez 
  affiliation: Institute for Computational Biomedicine, Heidelberg University 
   
# package: DoRothEA
output:
  BiocStyle::html_document
bibliography: Mybib.bib
abstract: |
  This vignette describes how to infer Transcription Factor activity from 
  scRNA-seq data by running Viper on DoRothEA's curated regulons. 
license: GNU-GLPv3, please check http://www.gnu.org/licenses/
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

**DoRothEA** is a comprehensive resource containing a curated collection of 
transcription factors (TFs) and its transcriptional targets. The set of genes
regulated by a specific transcription factor is known as regulon. DoRothEA's 
regulons were gathered from different types of evidence. Each TF-target 
interaction is defined by a confidence level based on the number of 
supporting evidence. The confidence levels ranges from A (highest confidence) 
to E (lowest confidence) [@Garcia-Alonso24072019]. 

**DoRothEA** regulons are usually coupled with the statistical method **VIPER** 
[@Alvarez2016]. In this context, TF activities are computed based on the mRNA 
expression levels of its targets. We therefore can consider TF activity as a 
proxy of a given transcriptional state [@dugourd2019]. 

Recently, @Holland2019 evaluated the performance of **DoRothEA** in 
combination with **VIPER** when applied to scRNA-seq data. They showed that, in 
spite of the current limitations of scRNA-seq technologies, their approach can 
provide meaningful results in this context. Indeed, this vignette shows an 
example on how to apply **DoRothEA** regulons coupled with **VIPER** in a well 
known single-cell dataset. 

# Installation

First of all, you need a current version of R (http://www.r-project.org). 
`r Biocpkg("DoRothEA")` is a freely available annotation package deposited on 
http://bioconductor.org/ and https://github.com/saezlab/DoRothEA. 

You can install it by running the following commands on an R console:

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DoRothEA")
```

We also load here the packages and functions required  to run this script. 

```{r, message = FALSE}
## We load the different required packages
library(dplyr)
library(Seurat)
library(ggplot2)
library(tidyr)
library(readr)
library(pheatmap)
library(viper)
library(parallel)
library(tibble)
library(purrr)
library(furrr)
library(umap)

options(future.globals.maxSize= 6291456000) # corresponds to 6 GB

#' VIPER wrapper
#'
#' This function is a convenient wrapper to run the 
#' \code{\link[=viper]{viper::viper()}} function to apply it on single cell 
#' data
#' 
#' @import Seurat
#' @import SingleCellExperiment
#' @import viper
#' @param InputObject An expression matrix with genes (HGNC symbol) in rows 
#' and samples in columns. It also accepts \code{Seurat} and 
#' \code{SingleCellExperiment} objects from which its normalised expresion 
#' matrix will be extracted 
#' @param regulon: Object of class regulon. Check  
#' \code{\link[=viper]{viper::viper()}} for further information
#' @param options A list of named options to pass to
#' \code{\link[=viper]{viper::viper()}} such as \code{minsize} or 
#' \code{method}. These options should not include \code{eset} or 
#' \code{regulon}.
#'
#' @return A matrix containing the activity of the different TFs provided in 
#' the regulon object.

sc_viper = function(InputObject, regulon, options=list()) {
    
    if (class(InputObject) == "Seurat"){
        expr <- InputObject[["RNA"]]@data
    } else {
        if(class(InputObject) == "SingleCellExperiment"){
            expr <- normcounts(InputObject)
        } else {
            expr <- InputObject
        }
    }
    
    emat <- as.matrix(expr)
    viper_res  <- do.call(viper,c(list(eset = emat,regulon = regulon),options))

    return(viper_res)
}

#' df2regulon: Function to group DoRothEA regulons
#'
#' This function takes a data frame containing the TF-target interactions
#' from DoRothEA and returns its associated regulons. 
#'
#' @import viper
#' @import dplyr
#' @import purrr
#' @param df A data frame containing the TF-target interactions from DoRothEA,
#' as stored in https://github.com/saezlab/ConservedFootprints/tree/master/data
#'
#' @return Object of class regulon. Check  
#' \code{\link[=viper]{viper::viper()}} for further information

df2regulon = function(df) {
    regulon = df %>%
        split(.$tf) %>%
        map(function(dat) {
            tf = dat %>% distinct(tf) %>% pull()
            targets = setNames(dat$mor, dat$target)
            likelihood = dat$likelihood
            list(tfmode =targets, likelihood = likelihood)
        })
    return(regulon)
}
```

# Example of usage

In the following paragraphs, we provide examples describing how to run **VIPER** 
on **DoRothEA** regulons in a scRNA-seq dataset. In particular, we use the 
`r CRANpkg("Seurat")` toolkit for single cell genomics [@Stuart2019]. For the 
sake of simplicity, we follow the example provided in the following 
`r CRANpkg("Seurat")` vignette:  

https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html 

The dataset contains 2,700 Peripheral Blood Mononuclear Cells (PBMC) that were 
sequenced on the Illumina NextSeq 500. This dataset is freely available in 
10X Genomics: 

https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz

```{r, eval=FALSE}
## Load the PBMC dataset
pbmc.data <- 
    Read10X(data.dir = "../data/pbmc3k/filtered_gene_bc_matrices/hg19/")
## Initialize the Seurat object with the raw (non-normalized data).
pbmc <- 
    CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, 
    min.features = 200)
```

```{r, eval=TRUE , include=FALSE}
load(file="/home/alvaldeolivas/Desktop/Trabajo/Heidelberg/ProgenySingleCellVignette/data/SeuratObject.RData", verbose = FALSE)
```

## Pre-processing, normalization and identification of highly variable features

We follow the standard pre-processing steps as described in the aforementioned 
`r CRANpkg("Seurat")` vignette before going deeper into the data analysis. 
These steps carry out the selection and filtration of cells based on quality
control metrics, the data normalization and scaling, and the detection of highly
variable features (see https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html). 

```{r, message=FALSE}
## Identification of mithocondrial genes
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

## Filtering cells following standard QC criteria.
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & 
    percent.mt < 5)

## Normalizing the data
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", 
    scale.factor = 10000)

pbmc <- NormalizeData(pbmc)

## Identify the 2000 most highly variable genes
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

## In addition we scale the data
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

## Clustering cells

One of the most relevant steps in scRNA-seq data analysis is clustering. Cells
are grouped based on the similarity of their transcriptomic profiles. We first 
apply the `r CRANpkg("Seurat")` v3 classical approach as described in their 
aforementioned vignette. We visualize the cell clusters using UMAP: 

```{r, message=FALSE, warning=FALSE}
pbmc <- 
    RunPCA(pbmc, features = VariableFeatures(object = pbmc), verbose = FALSE)
pbmc <- FindNeighbors(pbmc, dims = 1:10, verbose = FALSE)
pbmc <- FindClusters(pbmc, resolution = 0.5, verbose = FALSE)
pbmc <- RunUMAP(pbmc, dims = 1:10,  umap.method = 'umap-learn',  
    metric='correlation')
```

```{r}
DimPlot(pbmc, reduction = "umap")
```

@Holland2019 showed that clustering the cells based on their TF activity 
profiles can also be very interesting. Indeed, clustering cells using TF 
activity computed with **VIPER** and **DoRothEA** performs better than 
using the expression level of the same TFs. In addition, it brings 
complementary information to the clusters based on transcriptomics profiles.

Here, we first run **VIPER** on **DoRothEA's** regulons to obtain TFs activity.

```{r, message=FALSE}
## We read Dorothea Regulons for Human:
dorothea_regulon_human <- 
    read_csv("https://raw.githubusercontent.com/saezlab/ConservedFootprints/master/data/dorothea_benchmark/regulons/dorothea_regulon_human_v1.csv")

## We obtain the regulons based on interactions with confidence level A, B and C
regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) %>%
    df2regulon()

## We compute Viper Scores 
viper_scores <- sc_viper(pbmc, regulon, options = list(method = "scale", 
    minsize = 4, eset.filter = FALSE, cores = 1, verbose = FALSE))

## We transform Vipre scores into a data frame to better handling the resutls
viper_scores_df <- 
    data.frame(viper_scores,check.names = FALSE, stringsAsFactors = FALSE) %>% 
    rownames_to_column("tf") %>%
    data.frame(row.names = 1,stringsAsFactors = FALSE,check.names = FALSE) %>%
    t() 
```

We then apply `r CRANpkg("Seurat")` to cluster the cells following the same 
protocol than above.


```{r, message=FALSE}
## We compute the Nearest Neighbours to perform cluster
Neighbours <- FindNeighbors(viper_scores_df)
Clusters <- FindClusters(Neighbours[[1]], verbose = FALSE)

## We prepare the clusters to plot the results.
Clusters <- Clusters %>% 
    rownames_to_column("Cell") %>% 
    dplyr::rename(CellType = res.0.8)

## We compute the umap coordinates
dorothea_umap <- umap::umap(viper_scores_df, method="umap-learn") 
dorothea_umap_coords <- dorothea_umap$layout %>%
    data.frame(stringsAsFactors = FALSE, check.names = TRUE) %>%
    rownames_to_column("Cell") %>%
    dplyr::rename(UMAP_1 = X1, UMAP_2 = X2) %>%
    inner_join(Clusters)

```

```{r}
ggplot(dorothea_umap_coords, aes(UMAP_1, UMAP_2, color = CellType)) +
    geom_point() + theme_bw()

```

## TF activity per cell population

Following again `r CRANpkg("Seurat")` protocol, we next find the markers 
that help to identify and discriminate the different cell popualations 
present in the dataset under study.

```{r, message = FALSE}
## Finding differentially expressed features (cluster biomarkers)
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, 
    logfc.threshold = 0.25, verbose = FALSE)

## Assigning cell type identity to clusters
new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T", 
    "FCGR3A+ Mono", "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
```

We plot again the clusters along with their cell type labels

```{r, message = FALSE, warning = FALSE}
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

Finally, we characterice the different cell populations based on their 
TF activities thanks to the previously computed **VIPER** scores on 
**DoRothEA's** regulons. 

```{r, message = FALSE}
## We create a data frame containing the cells and their clusters
CellsClusters <- data.frame(Cell = names(Idents(pbmc)), 
                            CellType = as.character(Idents(pbmc)),
                            stringsAsFactors = FALSE)

## We create a data frame with the Viper score per cell and its clusters
viper_scores_clusters <- viper_scores_df  %>%
    data.frame() %>% 
    rownames_to_column("Cell") %>%
    gather(Tf, Activity, -Cell) %>%
    inner_join(CellsClusters)

## We summarize the Viper scores by cellpopulation
summarized_viper_scores <- viper_scores_clusters %>% 
  group_by(Tf, CellType) %>%
  summarise(avg = mean(Activity),
            std = sd(Activity))

```

For visualization purposes, we select the 20 most variable TFs across clusters
according to our scores. 

```{r, message=FALSE}
## We select the 20 most variable TFs. (20*9 populations = 180)
highly_variable_tfs <- summarized_viper_scores %>%
  group_by(Tf) %>%
  mutate(var = var(avg))  %>%
  ungroup() %>%
  top_n(180, var) %>%
  distinct(Tf)

## We prepare the data for the plot
summarized_viper_scores_df <- summarized_viper_scores %>%
  semi_join(highly_variable_tfs, by = "Tf") %>%
  dplyr::select(-std) %>%   
  spread(Tf, avg) %>%
  data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
```

```{r}
paletteLength = 100
myColor = colorRampPalette(c("Darkblue", "white","red"))(paletteLength)

viperBreaks <- c(seq(min(summarized_viper_scores_df), 0, 
                     length.out=ceiling(paletteLength/2) + 1),
                 seq(max(summarized_viper_scores_df)/paletteLength, 
                     max(summarized_viper_scores_df), 
                     length.out=floor(paletteLength/2)))
viper_hmap <- pheatmap(t(summarized_viper_scores_df),fontsize=14, 
                       fontsize_row = 10, 
                       color=myColor, breaks = viperBreaks, 
                       main = "DoRothEA (ABC)", angle_col = 45,
                       treeheight_col = 0,  border_color = NA) 
```


# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References
